Code:

Prepare genome:

```bash
wget https://ftp.ncbi.nlm.nih.gov/genomes/all/GCA/905/404/135/GCA_905404135.1_ilHesComm1.1/GCA_905404135.1_ilHesComm1.1_genomic.fna.gz
gunzip GCA_905404135.1_ilHesComm1.1_genomic.fna.gz
grep ">" GCA_905404135.1_ilHesComm1.1_genomic.fna | sed 's/>//'
nano W_id.txt
nano extra_scaffolds.txt
seqkit grep -v -n -f W_id.txt GCA_905404135.1_ilHesComm1.1_genomic.fna > GCA_905404135.1_ilHesComm1.1_genomic_noW.fa
samtools faidx GCA_905404135.1_ilHesComm1.1_genomic_noW.fa
bwa-mem2 index GCA_905404135.1_ilHesComm1.1_genomic_noW.fa
seqkit grep -v -n -f extra_scaffolds.txt GCA_905404135.1_ilHesComm1.1_genomic_noW.fa > Hcomma_genome.fa
samtools faidx Hcomma_genome.fa

awk '{ printf("%s\t0\t%s\n", $1, $2); }' Hcomma_genome.fa.fai > Hcomma_nuclear_chromosomes.bed

```

Remove lane number with this?

```sed
sed 's/_L00[0-9]//'
```


Map reads
```bash
bwa-mem2 mem -t 32 GCA_905404135.1_ilHesComm1.1_genomic_noW.fa HC-19-2016-30.veladapt.clean_R1.fastq.gz HC-19-2016-30.veladapt.clean_R2.fastq.gz | samtools sort -o HC-19-2016-30.sort.bam
```

Process bams:

```bash
java -jar ~/apps/picard/build/libs/picard.jar AddOrReplaceReadGroups \
-I HC-19-2016-30.sort.bam \
-O "HC-19-2016-30.RG.bam" \
-RGID G3modc \
-RGLB mod01 \
-RGPL ILLUMINA -RGPU unit1 \
-RGSM HC-19-2016-30

samtools view -@ 32 -b -f 3 -F 2828 -q 20 "HC-19-2016-30.RG.bam" -o "HC-19-2016-30.filtered.bam"

samtools sort -@ 32 -n "HC-19-2016-30.filtered.bam" -o "HC-19-2016-30.sorted.n.bam"

samtools fixmate -@ 32 -m "HC-19-2016-30.sorted.n.bam" "HC-19-2016-30.fixmate.bam"

samtools sort -@ 32 "HC-19-2016-30.fixmate.bam" -o "HC-19-2016-30.sorted.p.bam"

samtools markdup -r -@ 32 "HC-19-2016-30.sorted.p.bam" "HC-19-2016-30.markdup.bam"

samtools view -L Hcomma_nuclear_chromosomes.bed -o HC-19-2016-30.reduced.bam HC-19-2016-30.markdup.bam

samtools index HC-19-2016-30.reduced.bam
```

add or replace RGs:

```bash
gatk3 \
-T RealignerTargetCreator \
-R Hcomma_genome.fa \
-o "HC-19-2016-30.intervals" \
-I "HC-19-2016-30.reduced.bam"

gatk3 \
-T IndelRealigner \
-R Hcomma_genome.fa \
-targetIntervals "HC-19-2016-30.intervals" \
-I HC-19-2016-30.reduced.bam \
-o "HC-19-2016-30.realn.bam"
```
