# Prepare the reference genome

DToL have currently released 18/20 reference genomes, 17 of which have been annotated. Before mapping can begin, these must be downloaded, processed to remove unwanted contigs and indexed.

Genome files can be accessed from the ncbi genome page by clicking on the three dots under "Actions" and then "See more files on FTP". The link for ending in "genomic.fna.gz" should be downloaded and unzipped
```bash
wget https://ftp.ncbi.nlm.nih.gov/genomes/all/GCA/902/806/685/GCA_902806685.2_iAphHyp1.2/GCA_902806685.2_iAphHyp1.2_genomic.fna.gz
gunzip GCA_902806685.2_iAphHyp1.2_genomic.fna.gz
```
A list of all contigs can then be created using grep:
```bash
grep ">" GCA_902806685.2_iAphHyp1.2_genomic.fna | sed 's/>//'
```
Any additional unplaced contigs and the mtDNA should then be copied and pasted into a new text file. If a W chromosome is present, it should be placed into a separate file to be removed
```bash
nano Ahyperantus_excluded_contigs.txt
nano Hcomma_W.txt
```
The W should then be removed
```bash
seqkit grep -v -n -f Hcomma_W.txt GCA_905404135.1_ilHesComm1.1_genomic.fna > GCA_905404135.1_ilHesComm1.1_genomic_noW.fa
```
The genome containing autosomes, Z, mtDNA and unplaced contigs (everything except W) should then be indexed for mapping and initial processing
```bash
samtools faidx GCA_902806685.2_iAphHyp1.2_genomic.fna
bwa-mem2 index GCA_902806685.2_iAphHyp1.2_genomic.fna
```
The genome should then be filtered to remove the unplaced contigs and mtDNA
```bash
seqkit grep -v -n -f Ahyperantus_excluded_contigs.txt GCA_902806685.2_iAphHyp1.2_genomic.fna > Ahyperantus_genome.fa
```
Finally, the reduced fasta file should be indexed for GATK
```bash
samtools faidx Ahyperantus_genome.fa
picard CreateSequenceDictionary -R Ahyperantus_genome.fa -O Ahyperantus_genome.dict
```
