# Prepare the reference genome

DToL have currently released 18/20 reference genomes, 17 of which have been annotated. Before mapping can begin, these must be downloaded, processed to remove unwanted contigs and indexed.

Genome files can be accessed from the ncbi genome page by clicking on the three dots under "Actions" and then "See more files on FTP". The link for ending in "genomic.fna.gz" should be downloaded and unzipped
```bash
wget $LINK
gunzip $REF.fna.gz
```
A list of all contigs can then be created using grep:
```bash
grep ">" $REF.fna | sed 's/>//'
```
Any additional unplaced contigs and the mtDNA should then be copied and pasted into a new text file. If a W chromosome is present, its name should be recorded
```bash
nano extra_scaffolds.txt
```
The W should then be removed
```bash
seqkit grep -v -p $W_CHR $REF.fna > $REF_noW.fa
```
The genome containing autosomes, Z, mtDNA and unplaced contigs (everything except W) should then be indexed for mapping and initial processing
```bash
samtools faidx $REF_noW.fa
bwa-mem2 index $REF_noW.fa
```
The genome should then be filtered to remove the unplaced contigs and mtDNA
```bash
seqkit grep -v -n -f extra_scaffolds.txt $REF_noW.fa > $SPECIES_genome.fa
```
Finally, the reduced fasta file should be indexed for GATK
```bash
samtools faidx $SPECIES_genome.fa
picard CreateSequenceDictionary -R $SPECIES_genome.fa -O $SPECIES_genome.dict
```
