# Remove adaptors and map reads

## Adaptor trimming

Currently, I am using read files that have had their adaptors trimmed by Flora. Since I may need to refer back to her script, such as if I find that the adaptors were not removed properly, 
I can do so using her script [Remove adaptors.sh](https://github.com/mmerkin/Velocity/blob/main/scripts/Remove_adaptors.sh). Briefly, she created a list of all the adaptor sequences using Carl's library prep spreadsheet and then used
[Trimmomatic](https://github.com/usadellab/Trimmomatic) to remove them, as well as removing bases at the ends of reads with a quality below 20.

## Mapping

Although Flora had mapped the reads herself, I found that the indel realignment step would fail unless I remapped them; I suspect this is due to the way she removed the mitochondria and 
W chromosome sequences causing issues with indexing. Therefore, new alignment files should be created with the script [Map_reads.sh](https://github.com/mmerkin/Velocity/blob/main/scripts/Map_reads.sh) in the conda environment mem2. 
Before running, the variables at the top must be assigned values. Here is an example for the Ringlet modern core samples:
```bash
REF=~/velocity/Aphantopus_hyperantus/reference/Ahyperantus_genome.fa  # Path to reference genome
datapath=/pub61/florat/velocity/aphantopus_hyperantus/modern_data/00_raw_reads_modc/Samples  # Path to reads
output=~/velocity/Aphantopus_hyperantus/modc_mapping  # Path to output directory
threads=32  # Number of threads to use in mapping
```
