# Realignment around indels.

Most modern variant calling pipelines do not require an indel realignment step, so it is difficult to find a modern tool to perform it. It can still be acheived with GATK3.8, but the commands were
not carried over to GATK4. However, GATK3 requires a local installation, so one must first download the file and activate it in the conda environment GATK38. Alternatively, the correct version
of the package can be downloaded [here](https://anaconda.org/bioconda/gatk)

```bash
wget https://console.cloud.google.com/storage/browser/_details/gatk-software/package-archive/gatk/GenomeAnalysisTK-3.8-1-0-gf15c1c3ef.tar.bz2

tar -xvjf  GenomeAnalysisTK-3.8-1-0-gf15c1c3ef.tar.bz2

cd GenomeAnalysisTK-3.8-1-0-gf15c1c3ef/

gatk-register GenomeAnalysisTK.jar
```

If the website has been taken down, search for a new link on the forums.

Finally, run the script [Indel_realignment.sh](https://github.com/mmerkin/Velocity/blob/main/scripts/Indel_realignment.sh), setting the variables at the top. This script will produce a log file
of all the individuals where the realignment didn't work

Example variables for ringlet modern core samples:

```bash
REF=~/velocity/Aphantopus_hyperantus/reference/Ahyperantus_genome.fa
datapath=~/velocity/Aphantopus_hyperantus/modc_processing
output=~/velocity/Aphantopus_hyperantus/modc_realign
fail_log=Ah_modc_realign_fails.txt
```
