#  Input file creation

Most population genetics stats can be calculated from genotype likelihood files, although these might require a vcf or sfs intermediate file.

It is recommended to make a backup of the recalibration files before proceeding as an incorrect input can cause them to be deleted.
```bash
cp modc_final modc_backup
```

## GLF

ATLAS can be used to call genotype likelihoods in parallel:

```bash
parallel -j 10 --memfree 400G --retries 10 --joblog AH_glf_modc.log \
~/apps/atlas/build/atlas GLF --RGInfo {.}_RGInfo.json \
 --bam {} ::: ~/velocity/Aphantopus_hyperantus/modc_final/*/*.bam
```

This uses much less memory than estimateErrors, so it can be used on more samples at once and will finish much faster. The output glf.gz file can then be viewed using:

```bash
atlas printGLF --glf AH-02-2019-42_final.glf.gz | less
```

## majorMinor

Variants can be called based on the entire sample using majorMinor. Note that it is much better to use GLFs when possible, but some tools require a vcf as input. Since the entire sample
is used as an input, this process does not need to be run in parallel.

First a text file needs to be created with the paths to all the glf.gz files on a new line. The sample names can be changed by creating a list of all the names separated by commas and no spaces,
which is stored as a variable "names" here.

```bash
ls */*.glf.gz > AH_mode_glfs.txt
names=$(ls -l | grep "^d" | awk '{ printf "%s,", $9 }' | sed 's/,$//')

atlas majorMinor --glf AH_modc_glfs.txt --maxThreads 32 --minMAF 0.005 --phredLik --sampleNames $names \
--fasta ~/velocity/Aphantopus_hyperantus/reference/Ahyperantus_genome.fa --out AH_mode_majmin
```

## SFS

This is still a WIP
