# GONE

GONe is a shell script tool that allows for the inference of *N<sub>e</sub>* from a population across many generations, although it is most accurate between 10 and 200. The script can be downloaded 
from its [github page](https://github.com/esrud/GONE):

```bash
git clone https://github.com/esrud/GONE.git
```

## Filtering

GONe requires a plink ped file input, which can be generated from the atlas major minor vcf. However, there is a maximum limit of 10 million SNPs and 1 million per chromosome, 
so initial filtering should be applied.

First, the vcf is filtered to remove the Z chromosome and low quality calls. If there are still too many variants, they can be subsampled using the vcflib tool vcfrandomsample:

```bash
vcftools 
bcftools view -e 'QUAL<30' -O b -o AHmode_GQ30.bcf AH_mode_noZ.recode.vcf.gz
bcftools view AHmode_renamed.vcf | vcfrandomsample -r 0.1 > AHmode_subset.vcf

#View the number of SNPs on each chromosome:
for i in {1..$AUTOSOME_NUMBER}; do echo $i && grep $i AHmodc_renamed.vcf | wc -l; done
```

Next, the vcf needs to be reformated for input into plink by replacing the chromosome names with numbers and adding SNP ids:

```bash
#Create a table of each chromosome name and their corresponding number
awk '{ printf "%s\t%d\n", $1, NR }' < $REF.fa.fai > $SPECIES_chrs.txt
#Rename chromosomes and replace missing ids (+) with the chromosome name, position and each allele separated by an underscore.
bcftools annotate --rename-chrs AH_chrs.txt --set-id +'%CHROM\_%POS\_%REF\_%ALT' -Ov -o AHmode_renamed.vcf AHmode_GQ30.bcf
```
