# GONE

GONe is a shell script tool that allows for the inference of *N<sub>e</sub>* from a population across many generations, although it is most accurate between 10 and 200. The script can be downloaded 
from its [github page](https://github.com/esrud/GONE):

```bash
git clone https://github.com/esrud/GONE.git
```

## Filtering

GONe requires a plink ped file input, which can be generated from the atlas major minor vcf. However, there is a maximum limit of 10 million SNPs and 1 million per chromosome, 
so initial filtering should be applied.

First, the vcf is filtered to remove the Z chromosome and low quality calls. If there are still too many variants, they can be subsampled using the vcflib tool vcfrandomsample. Filetag is just
the name to call all the output files

```bash
vcftools --gzvcf $MAJMIN_VCF --not-chr $Z_CHROM --recode --recode-INFO-all --out $VCF_NOZ
bcftools view -e 'QUAL<30' -O b -o $FILETAG_QS30.bcf $VCF_NOZ.recode.vcf.gz
bcftools view $FILETAG_QS30.vcf | vcfrandomsample -r 0.1 > $FILETAG_subset.vcf

# View the number of SNPs on each chromosome:
for i in {1..$AUTOSOME_NUMBER}; do echo $i && grep $i $FILETAG_subset.vcf | wc -l; done
```

Next, the vcf needs to be reformated for input into plink by replacing the chromosome names with numbers and adding SNP ids:

```bash
# Create a table of each chromosome name and their corresponding number
awk ' NR<=$AUTOSOME_NUMBER { printf "%s\t%d\n", $1, NR }' < $REF.fa.fai > $SPECIES_chrs.txt
# Rename chromosomes and replace missing ids (+) with the chromosome name,
# its position and each allele separated by an underscore.
bcftools annotate --rename-chrs $SPECIES_chrs.txt --set-id +'%CHROM\_%POS\_%REF\_%ALT' \
-Ov -o $FILETAG_renamed.vcf $FILETAG_subset.vcf
```

It may be necessary to create a subset of only certain individuals, which can be acheived by creating a text file of each sample name on a new line and filtering with --keep. Note that the use of 
underscores and dashes can be inconsistent between the vcf and spreadsheets, so they may have to be renamed to match the vcf.

```bash
# Create file with sample names
nano $POP_list.txt
# Replace underscores with dashes
sed -i 's/_/-/g' < $POP_list.txt
# Filter the vcf
vcftools --vcf $FILETAG_RENAMED.vcf --keep $POP_list.txt --recode --out $FILETAG_$POP
```

The plink files can now be created for input into GONE, although they should first be modified by adding a -9 before the first allele (all of the columns before this number are ignored).

```bash
vcftools --vcf AHmode_subset.vcf --out $FILETAG --plink
awk '$6="-9"' $FILETAG.ped > $FILETAG.temp.txt
rm $FILETAG.ped
mv $FILETAG.temp.txt $FILETAG.ped
```

Finally, GONE can be run to produce an output file called Output_Ne_$FILETAG. The first line of this file should be deleted before opening it in R.

```bash
bash script_GONE.sh $FILETAG
tail -n +2 Output_Ne_$FILETAG
```
